domain/src/main/java/no/difi/meldingsutveksling/arkivmelding/ArkivmeldingFactory.java

```diff
@@ -0,0 +1,162 @@
+        ofNullable(mt.getNoarksak().getSaSeknr()).map(BigInteger::new).ifPresent(sm::setSakssekvensnummer);
+        ofNullable(mt.getNoarksak().getSaAnsvinit()).ifPresent(sm::setSaksansvarlig);
+        ofNullable(mt.getNoarksak().getSaAdmkort()).ifPresent(sm::setAdministrativEnhet);
+        ofNullable(mt.getNoarksak().getSaOfftittel()).ifPresent(sm::setOffentligTittel);
+
+        if (mt.getJournpost() == null) {
+            throw new MeldingsUtvekslingRuntimeException(format("No Journpost in MeldingType for message %s, aborting conversion", putMessage.getConversationId()));
+        }
+        Journalpost jp = amOf.createJournalpost();
+        ofNullable(mt.getJournpost().getJpJaar()).map(BigInteger::new).ifPresent(jp::setJournalaar);
+        ofNullable(mt.getJournpost().getJpSeknr()).map(BigInteger::new).ifPresent(jp::setJournalsekvensnummer);
+        ofNullable(mt.getJournpost().getJpJpostnr()).map(BigInteger::new).ifPresent(jp::setJournalpostnummer);
+        ofNullable(mt.getJournpost().getJpNdoktype()).map(JournalposttypeMapper::getArkivmeldingType).ifPresent(jp::setJournalposttype);
+        Skjerming skjerming = amOf.createSkjerming();
+        ofNullable(mt.getJournpost().getJpUoff()).ifPresent(skjerming::setSkjermingshjemmel);
+        sm.setSkjerming(skjerming);
+
+        // expecting date in format yyyy-MM-dd
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        Optional<String> jpDato = ofNullable(mt.getJournpost().getJpJdato());
+        if (jpDato.isPresent()) {
+            LocalDate localDate = LocalDate.parse(jpDato.get(), formatter);
+            GregorianCalendar gcal = GregorianCalendar.from(localDate.atStartOfDay(ZoneId.systemDefault()));
+            XMLGregorianCalendar xgcal = null;
+            try {
+                xgcal = DatatypeFactory.newInstance().newXMLGregorianCalendar(gcal);
+            } catch (DatatypeConfigurationException e) {
+                log.error("Could not convert JpJdato in message {}", putMessage.getConversationId(), e);
+            }
+            ofNullable(xgcal).ifPresent(jp::setJournaldato);
+        }
+        Optional<String> jpDokdato = ofNullable(mt.getJournpost().getJpDokdato());
+        if (jpDokdato.isPresent()) {
+            LocalDate localDate = LocalDate.parse(jpDokdato.get(), formatter);
+            GregorianCalendar gcal = GregorianCalendar.from(localDate.atStartOfDay(ZoneId.systemDefault()));
+            XMLGregorianCalendar xgcal = null;
+            try {
+                xgcal = DatatypeFactory.newInstance().newXMLGregorianCalendar(gcal);
+            } catch (DatatypeConfigurationException e) {
+                log.error("Could not convert JpJdato in message {}", putMessage.getConversationId(), e);
+            }
+            ofNullable(xgcal).ifPresent(jp::setDokumentetsDato);
+        }
+
+        mt.getJournpost().getAvsmot().forEach(a -> {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce this lambda expression number of lines from 28 to at most 20. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-app01.difi.local:9000/coding_rules#rule_key=squid%3AS1188)